Add custom properties support + unit-tests, feature-tests

michael-koeller opened the pull request:
Hi @scanny,

I patched the implementation of @renejsum into a new fork and added the missing test cases.
Along the way I also added proper type handling for string/int/bool properties.

Regards,
Michael

renejsum commented:
Awesome :-)

> On 23 Nov 2018, at 11.47, Michael Köller <notifications@github.com> wrote:
> 
> Hi @scanny <https://github.com/scanny>,
> 
> I patched the implementation of @renejsum <https://github.com/renejsum> into a new fork and added the missing test cases.
> Along the way I also added proper type handling for string/int/bool properties.
> 
> Regards,
> Michael
> 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/python-openxml/python-docx/pull/580 <https://github.com/python-openxml/python-docx/pull/580>
> Commit Summary
> 
> Added customprops
> add tests, fix type handling
> File Changes
> 
> M docx/__init__.py <https://github.com/python-openxml/python-docx/pull/580/files#diff-0> (2)
> M docx/document.py <https://github.com/python-openxml/python-docx/pull/580/files#diff-1> (8)
> M docx/opc/constants.py <https://github.com/python-openxml/python-docx/pull/580/files#diff-2> (3)
> A docx/opc/customprops.py <https://github.com/python-openxml/python-docx/pull/580/files#diff-3> (72)
> M docx/opc/package.py <https://github.com/python-openxml/python-docx/pull/580/files#diff-4> (22)
> A docx/opc/parts/customprops.py <https://github.com/python-openxml/python-docx/pull/580/files#diff-5> (74)
> M docx/oxml/__init__.py <https://github.com/python-openxml/python-docx/pull/580/files#diff-6> (11)
> A docx/oxml/customprops.py <https://github.com/python-openxml/python-docx/pull/580/files#diff-7> (156)
> M docx/oxml/ns.py <https://github.com/python-openxml/python-docx/pull/580/files#diff-8> (1)
> M docx/parts/document.py <https://github.com/python-openxml/python-docx/pull/580/files#diff-9> (8)
> A features/doc-customprops.feature <https://github.com/python-openxml/python-docx/pull/580/files#diff-10> (28)
> A features/steps/customprops.py <https://github.com/python-openxml/python-docx/pull/580/files#diff-11> (89)
> A features/steps/test_files/doc-customprops.docx <https://github.com/python-openxml/python-docx/pull/580/files#diff-12> (0)
> A features/steps/test_files/doc-no-customprops.docx <https://github.com/python-openxml/python-docx/pull/580/files#diff-13> (0)
> A tests/opc/parts/test_customprops.py <https://github.com/python-openxml/python-docx/pull/580/files#diff-14> (51)
> A tests/opc/test_customprops.py <https://github.com/python-openxml/python-docx/pull/580/files#diff-15> (90)
> Patch Links:
> 
> https://github.com/python-openxml/python-docx/pull/580.patch <https://github.com/python-openxml/python-docx/pull/580.patch>
> https://github.com/python-openxml/python-docx/pull/580.diff <https://github.com/python-openxml/python-docx/pull/580.diff>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/python-openxml/python-docx/pull/580>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AAjLh1IGAUSMvRewQ74xh6gca4bm6ci-ks5ux9JbgaJpZM4Ywcsn>.
>

Ricyteach commented:
I haven't reviewed the PR but wanted to say thanks, @michael-koeller!

jekozyra commented:
@scanny any chance of this ever making it into a release? It would save me a lot of time.

BastienFaure commented:
Merging this pull-request would we fantastic :) Thanks to all you guys for your job !

geobeo commented:
This would basically be the one and only reason for me to use python-docx (a lot). I consider this a killer feature :P Please merge (or give someone write access to merge :P).

zebulon501 commented on docx/opc/package.py:
```diff
@@ -41,6 +42,14 @@ def core_properties(self):
+    def custom_properties(self):
+        """
+        |CustomProperties| object providing read/write access to the Dublin
+        Core properties for this document.
```
Custom properties, not Dublin Core.

michael-koeller replied:
Fixed

sanzoghenzo commented on docx/opc/customprops.py:
```diff
@@ -0,0 +1,71 @@
+"""
+The :mod:`pptx.packaging` module coheres around the concerns of reading and
+writing presentations to and from a .pptx file.
+"""
```
Fix docstring

michael-koeller replied:
Fixed.

nozokada commented:
I'm so excited for this to be merged to master. Thank you for the work.

Ricyteach commented:
> I'm so excited for this to be merged to master. Thank you for the work.

Seconded. I intend to use the stink out of this feature. THANK YOU!!!

paweljasinski commented:
Can this be merged?
If not, what is outstanding?
Thanks in advance

HansBambel commented:
Would be nice to have this feature merged in :)

mohitmathew commented:
Can we get this merged please ? this has been open for a very long time now.

hugoe commented:
Yes, I need it too, please merge.

boomesq commented:
I need it as well.

lsaint commented:
It's been four years. Merge, please.

Apteryks commented:
Hi, and thanks for this work!  

It looks rather complete, with the acceptance tests and unit tests already in!  I've taken the freedom to rebase it and get rid of the merge commits; I believe I've preserved the correct authors and the tests still passes; @michael-koeller could you hard reset your branch to https://github.com/Apteryks/python-docx/tree/add-support-for-custom-properties, after verifying that it looks alright?

After which we can look into some details.

sanzoghenzo commented:
Hi, there is still a fix to do in the docstring that refers to pptx package instead of docx. See my review above.

I'm glad you're taking the time and effort to keep the ball rolling!

michael-koeller commented:
@Apteryks: Nice to hear, that things might get some traction after laying sleeping for so long. 👍 

I just did a rebase on my pull request branch and I also added a separate feature branch in my repo wich I didn't do originally (don't know any more why).

Apteryks commented:
Hi @michael-koeller; for tidiness, could you please split the acceptance tests in a first commit, and everything else into a 2nd commit?  Basically the way I did it here: https://github.com/Apteryks/python-docx/commits/add-support-for-custom-properties.

michael-koeller commented:
> for tidiness, could you please split the acceptance tests in a first commit, and everything else into a 2nd commit?

Yes, sure. Done.

Apteryks reviewed (changes requested):
Hi!  This looks good; I added a few minor comments.

Testing it though, I seem to trigger a hang this way:
```
from docx import Document
doc = Document('./features/steps/test_files/doc-customprops.docx')
list(doc.custom_properties)
```

Is this expected?

Thanks.

Apteryks commented:
@michael-koeller One last thing: it'd be nice to add documentation for the new feature; it seems it'd go here: https://github.com/python-openxml/python-docx/blob/master/docs/api/document.rst, next to the CoreProperties section.  A usage example would also be neat in https://raw.githubusercontent.com/python-openxml/python-docx/master/docs/index.rst, demonstrating how to access or set custom properties on a document. Thanks again!

mohitmathew commented:
Almost there team can we push this through 👍 . Thanks for everyone's effort ..

dmuensterer commented:
Hi, thanks for your effort.
In the meantime I’ve been working on supporting app properties as well. I.e. the company or application.
I’ve essentially cloned the core property classes and changed the necessary bits. However I am not able to come by an error that’s thrown when trying to create my AppProperty element.
@Apteryks @michael-koeller would either of you be willing to have a look? 
I think after supporting all types of properties python-docx would be significantly more convenient!

danyi-david commented on docx/opc/parts/customprops.py:
```diff
@@ -0,0 +1,74 @@
+        values for its base properties.
+        """
+        custom_properties_part = cls._new(package)
+        custom_properties = custom_properties_part.custom_properties
```
custom_properties is never used

michael-koeller replied:
Resolved.

celkas commented:
@michael-koeller anything I can help to get this MR ready?

michael-koeller commented:
> @michael-koeller anything I can help to get this MR ready?

Sorry for the delay.
The adjustments also keep bubbling up and nagging on my task list. I hope to get them done soon.

michael-koeller commented on docx/oxml/customprops.py:
```diff
@@ -0,0 +1,156 @@
+        # strptime isn't smart enough to parse literal timezone offsets like
+        # '-07:30', so we have to do it ourselves
+        parseable_part = w3cdtf_str[:19]
+        offset_str = w3cdtf_str[19:]
```
String slicing out of bounds does not cause an error in python, but rather returns an empty sequence.
So this W3C date parsing method is robust in this respect.

michael-koeller commented on docx/opc/customprops.py:
```diff
@@ -0,0 +1,70 @@
+        if prop is not None :
+            # print(etree.tostring(prop, pretty_print=True))
+            elm = prop[0]
+            if elm.tag == '{%s}i4' % NS_VT:
```
f-string are available for python3.6+ only.
Since the rest of the project code still supports python 2.x, I'd rather not break that.

michael-koeller commented on docx/opc/customprops.py:
```diff
@@ -0,0 +1,70 @@
+            elm = etree.SubElement(prop, '{%s}%s' %(NS_VT, elmType), nsmap = {'vt':NS_VT} )
+            elm.text = value
+            prop.set("name", key)
+            prop.set("fmtid", "{D5CDD505-2E9C-101B-9397-08002B2CF9AE}")
```
Resolved.

michael-koeller commented on docx/opc/customprops.py:
```diff
@@ -0,0 +1,70 @@
+            prop.set("pid", "%s" % str(len(self._element) + 1))
+        else:
+            elm = prop[0]
+            if elm.tag == '{%s}i4' % NS_VT:
```
See above. f-strings are only available for python 3.6+.

michael-koeller commented on docx/oxml/__init__.py:
```diff
@@ -39,6 +39,14 @@ def register_element_cls(tag, cls):
 
+def register_element_cls_ns(tag, ns, cls):
+    """
+    Register *cls* to be constructed when the oxml parser encounters an
```
Resolved.

michael-koeller commented on docx/opc/customprops.py:
```diff
@@ -0,0 +1,71 @@
+    def __getitem__( self, item ):
+        prop = self.lookup(item)
+        if prop is not None :
+            # print(etree.tostring(prop, pretty_print=True))
```
Resolved.

michael-koeller commented on docx/oxml/__init__.py:
```diff
@@ -72,7 +80,10 @@ def OxmlElement(nsptag_str, attrs=None, nsdecls=None):
+from .customprops import CT_CustomProperties
+register_element_cls_ns('Properties', 'http://schemas.openxmlformats.org/officeDocument/2006/custom-properties', CT_CustomProperties)
+
+from .document import CT_Body, CT_Document
```
Resolved.

michael-koeller commented on docx/opc/customprops.py:
```diff
@@ -0,0 +1,70 @@
+            if elm.tag == '{%s}i4' % NS_VT:
+                try:
+                    return int(elm.text)
+                except:
```
Resolved.

michael-koeller commented on docx/oxml/customprops.py:
```diff
@@ -0,0 +1,156 @@
+            # These two require an explicit 'xsi:type="dcterms:W3CDTF"'
+            # attribute. The first and last line are a hack required to add
+            # the xsi namespace to the root element rather than each child
+            # element in which it is referenced
```
Resolved.

michael-koeller commented on docx/oxml/customprops.py:
```diff
@@ -0,0 +1,156 @@
+        """
+        value = str(value)
+        if len(value) > 255:
+            tmpl = (
```
Resolved.

michael-koeller commented on tests/opc/parts/test_customprops.py:
```diff
@@ -0,0 +1,51 @@
+    # fixtures ---------------------------------------------
+
+    @pytest.fixture
+    def customprops_fixture(self, element_, CustomProperties_):
```
Resolved.

michael-koeller commented on tests/opc/test_customprops.py:
```diff
@@ -0,0 +1,90 @@
+from docx.opc.customprops import CustomProperties
+from docx.oxml.customprops import CT_CustomProperties
+from docx.oxml import parse_xml
+from lxml import etree
```
Resolved.

michael-koeller commented on tests/opc/test_customprops.py:
```diff
@@ -0,0 +1,90 @@
+
+    # fixture components ---------------------------------------------
+
+    def customProperties(self, prop_name, str_type, str_value):
```
Resolved.

michael-koeller commented:
> > @michael-koeller anything I can help to get this MR ready?
> 
> Sorry for the delay. The adjustments also keep bubbling up and nagging on my task list. I hope to get them done soon.

OK, done. As far as I can see.

celkas commented:
Almost there team can we push this through 👍 . Thanks for everyone's effort ..

R00118189 commented:
@celkas what is this feature estimated merge date?

celkas commented:
@Apteryks I tested it on Windows. TL;DR same result that You have in https://github.com/python-openxml/python-docx/pull/580#pullrequestreview-1048746223

```
from docx import Document

print("-------------------------------------------")
print("opening doc with custom-props => works fine")
print("-------------------------------------------")
name = r"C:\Python\python-docx-master\features\steps\test_files\doc-customprops.docx"
doc = Document(name)
print(name)
print(doc)
print(doc.custom_properties)

print("-------------------------------------------")
print("access using known values => works fine")
print("-------------------------------------------")
known_propvals = (
    ("CustomPropBool", True),
    ("CustomPropInt", 13),
    ("CustomPropString", "Test String"),
)
for name, expected_value in known_propvals:
    print(name, doc.custom_properties[name])

print("-------------------------------------------")
print("access like a list or dict => fails")
print("-------------------------------------------")
i = 0
for prop in doc.custom_properties:
    if i > 10:
        break
    print(prop)
    i += 1
list(doc.custom_properties)

print("-------------------------------------------")
print("end")
print("-------------------------------------------")
```
results in
```
C:\...>python docx-test.py
-------------------------------------------
opening doc with custom-props => works fine
-------------------------------------------
C:\Python\python-docx-master\features\steps\test_files\doc-customprops.docx
<docx.document.Document object at 0x000001E784E8FF40>
<docx.opc.customprops.CustomProperties object at 0x000001E7870DBE20>
-------------------------------------------
access using known values => works fine
-------------------------------------------
CustomPropBool True
CustomPropInt 13
CustomPropString Test String
-------------------------------------------
access like a list or dict => fails
-------------------------------------------
None
None
None
None
None
None
None
None
None
None
None
Traceback (most recent call last):
  File "C:\...\docx-test.py", line 32, in <module>
    list(doc.custom_properties)
  File "C:\Python\Python310\lib\site-packages\docx\opc\customprops.py", line 26, in __getitem__
    prop = self.lookup(item)
  File "C:\Python\Python310\lib\site-packages\docx\opc\customprops.py", line 70, in lookup
    if child.get("name") == item:
KeyboardInterrupt
^C
C:\...>
```

Apteryks commented:
> @Apteryks I tested it on Windows. TL;DR same result that You have in [#580 (review)](https://github.com/python-openxml/python-docx/pull/580#pullrequestreview-1048746223)
Thanks for testing.  So I think this should be fixed before we poke the maintainer to have a look and hopefully merge; @michael-koeller, is this something you'd be interested looking into?  Giving a more dict-like interface to the properties (or storing them in a plain dict directly)

celkas commented:
@michael-koeller, easiest way I see is adding this code to `docx\opc\customprops.py`:
```
    def __iter__(self):
        return iter(self._element)
```
This would make it possible to iterate all custom props in a file.

One needs to be aware, though, that the access to a custom-prop is (then) still not straight forward:
```
for prop in doc.custom_properties:
    #print(prop) # does not work!
    print(type(prop), prop.items(), prop.get('name'))
```

michael-koeller commented:
I just added the iteration support on custom properties, and added the relevant test cases for this feature.

@celkas : Thanks for your input. My version is slightly different from yours. The iteration just returns the names of existing custom properties. To obtain the values, any client code may then explicitely access the getter method.

celkas commented:
@Apteryks, looks good to me. Tested the branch, everything OK.

jakeformico commented:
I might be a little out of my depth here, but I'm hoping someone can clear this up. With this, I was successfully able to get and update my custom property, which in my case was used for updating the existing header, but I noticed inside the actual word document nothing changed. The old header stayed the same. Then I checked, and it looks like I need to update the field even after I've updated the custom property itself.  Does this functionality also handle updating the fields associated with the custom properties?

R00118189 commented:
> I might be a little out of my depth here, but I'm hoping someone can clear this up. With this, I was successfully able to get and update my custom property, which in my case was used for updating the existing header, but I noticed inside the actual word document nothing changed. The old header stayed the same. Then I checked, and it looks like I need to update the field even after I've updated the custom property itself. Does this functionality also handle updating the fields associated with the custom properties?

I believe you have to refresh it manually in MSWord and save it to keep it displayed or set MSWord to update fields before printing *Click FILE > Options > Display, and under Printing options, select the check box for Update fields before printing.* [see here](https://support.microsoft.com/en-us/office/update-fields-7339a049-cb0d-4d5a-8679-97c20c643d4e)

R00118189 commented:
@michael-koeller, how do we remove or delete custom properties from <class 'docx.opc.customprops.CustomProperties'> object?

michael-koeller commented:
> @michael-koeller, how do we remove or delete custom properties from <class 'docx.opc.customprops.CustomProperties'> object?

Good question. That's not covered yet, as it seems.

michael-koeller commented:
> @michael-koeller, how do we remove or delete custom properties from <class 'docx.opc.customprops.CustomProperties'> object?

OK, should be working now. 😉 

I also added corresponding test cases.

R00118189 commented:
@michael-koeller do we delete by setting the value (that does not work, as properties stay with empty values) to None or is there a way to access the delete function and get rid of the property altogether?

michael-koeller commented:
> @michael-koeller do we delete by setting the value (that does not work, as properties stay with empty values) to None or is there a way to access the delete function and get rid of the property altogether?

Ah, sorry. I forgot to explain:

```python
    custom_properties = document.custom_properties
    # delete custom property "Foo"
    del custom_properties["Foo"]
```

The implementation is defensive. So if there's no custom property to the given key, nothing happens.

celkas commented:
What can be done to make this happen?

Apteryks commented:
> 



> ```python
>     custom_properties = document.custom_properties
>     # delete custom property "Foo"
>     del custom_properties["Foo"]
> ```

I'm wondering, is there a reason the properties can't be kept in a true Python `dict` object, instead of a look-alike class?  I think the API would be more intuitive if this was the case (we can simply document that the properties are exposed as a dictionary, the same way the environment variables are exposed as such via `os.environ`).  What do you think?

michael-koeller commented:
> I'm wondering, is there a reason the properties can't be kept in a true Python dict object, instead of a look-alike class?

That's already part of the original implementation by @renejsum.

The util class actually operates directly on the XML document tree. So while it might as well be possible to keep the custom properties in a plain dictionary, you would still need code to couple this dictionary instance to the XML elements in the tree.

Apteryks commented:
Hi everyone, and happy New Year!

I've tested this again, and the test suite passes (`============================= 1477 passed in 5.10s =============================`), and the implementation seems to work alright.

My preference would be to have a better emulated dict-like interface (where you could call .values() or .items()), but that can always be added later.

@scanny in case you haven't followed this discussion, this is about adding support for custom properties;  the change includes tests and has been reviewed and manually tested by myself and others.

R00118189 commented:
Lads, what is the roadmap for merging it into the repo?

Apteryks commented:
> Lads, what is the roadmap for merging it into the repo?

@scanny as the owner of the project has the last word; it seems they haven't had the time to catch up with this thread yet; I hope they do, as it seems to be a much wanted feature :-).

jzzsxm commented:
@michael-koeller and @Apteryks I appreciate all the hard work on this. I've just implemented this new functionality at work and it's a huge time saver. Very well done!

Kasimashi commented:
Any news about it ?

sHermanGriffiths commented:
@michael-koeller  @scanny, any updates?

celkas commented:
@michael-koeller it seems there are conflicts to be resolved? At least that is a new message appearing in my feed...

scanny commented:
Hi All, apologies for the long hiatus. This PR looks potentially viable. @michael-koeller are you still interested in driving this?

I have a few comments but no sense getting into that if it's gone stale. I would need a rebase given all the changes in getting from `0.8` to `1.0` and dropping the Python 2 support.

Let me know if you're still interested and we can see where to go next.

michael-koeller commented:
Hi Steve,

great to hear from you. And yes, I’m still interested in having this feature merged into the project.

I  hope that I find the time today to make a rebase on my MR.
If you have any further comments, please don’t hesitate to hand them to me.


> Am 20.10.2023 um 20:10 schrieb Steve Canny ***@***.***>:
> 
> 
> Hi All, apologies for the long hiatus. This PR looks potentially viable. @michael-koeller <https://github.com/michael-koeller> are you still interested in driving this?
> 
> I have a few comments but no sense getting into that if it's gone stale. I would need a rebase given all the changes in getting from 0.8 to 1.0 and dropping the Python 2 support.
> 
> Let me know if you're still interested and we can see where to go next.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/python-openxml/python-docx/pull/580#issuecomment-1773182613>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AALKVDLURDK276XLZAA5RJTYAK5ATAVCNFSM4GGBZMT2U5DIOJSWCZC7NNSXTN2JONZXKZKDN5WW2ZLOOQ5TCNZXGMYTQMRWGEZQ>.
> You are receiving this because you were mentioned.
>

michael-koeller commented:
Oops.

While working on a rebase for this MR, I accidentally closed this MR. Sorry for any inconveniences. 😞 

As a possible followup, I created a new MR on my recent branch: #1273
